### PR TITLE
Effect queue+projectile copy issues

### PR DESCRIPTION
**Pull request by bradallred**
This is all fallout from MSVC pointing out we aren't move compliant where we intend and need to be

not tested yet, pushing to check build status

### CODE REVIEW

**Comment by bradallred** [comment]
my sources tell me this appears to be a bug in gcc. we should try with something more recent than gcc7.

**Comment by lynxlynxlynx** [comment]
We can bump gcc on the bot and document a minimum version. Unfortunately the build is still failing on msvc as well, so there's more to be done.

**Comment by MarcelHB** [comment]
But maybe try to have GCC 8 running since that's the default on RPi OS. I'll have a fresh look in a minute.

**Comment by MarcelHB** [comment]
It's again some STL container ... What about doing it the hard-way in these cases? Like giving up `noexcept = default` and doing that stuff manually. It's a shit-load of work in some cases but compiler is happy then.

**Comment by bradallred** [comment]
It's an option to specify them manually, of course, but that is undesirable for a number of reasons. Its likely to propagate well beyond the current complaint and manually managing copy constructors becomes a place to introduce bugs when somebody adds a new field.

**Comment by bradallred** [comment]
GCC is back in the green here, we should do some testing before merging this tho

**Comment by MarcelHB** [comment]
Great. I can test building and playing around a bit with what looks related.

We want to try to fix MSVC in a different PR then?

**Comment by bradallred** [comment]
I'm still working on MSVC, there are indeed problems with at leas `WorldMap`

**Comment by bradallred** [comment]
Good, that last commit made MSVC shut about `WorldMap`

**Comment by bradallred** [comment]
ok, thats as much fixing as MSVC is gonna get for now. We still need #1542 to get it green.

File: gemrb/core/Projectile.cpp [hunk 1]
```diff
@@ -158,8 +152,10 @@ void Projectile::CreateOrientedAnimations(Animation **anims, const AnimationFact
 			break;
 		}
 		Animation* a = af->GetCycle( c );
-		anims[Cycle] = a;
 		if (!a) continue;
+		
+		anims[Cycle] = std::move(*a);
```
Comment (MarcelHB) [inline gemrb/core/Projectile.cpp:157]: When casting a spell, I get a crash:
```
Thread 1 received signal SIGSEGV, Segmentation fault.
0x00007ff8f382192b in GemRB::Animation::operator= (this=0x0) at D:/Sources/gemrb/gemrb/core/Animation.h:38
38      class GEM_EXPORT Animation {
(gdb) bt
#0  0x00007ff8f382192b in GemRB::Animation::operator= (this=0x0)
    at D:/Sources/gemrb/gemrb/core/Animation.h:38
#1  0x00007ff8f3749924 in GemRB::Projectile::CreateOrientedAnimations (
    this=0x25babc77a80, af=0x25babbe7330, Seq=0)
    at D:/Sources/gemrb/gemrb/core/Projectile.cpp:157
#2  0x00007ff8f37496a3 in GemRB::Projectile::CreateAnimations (
    this=0x25babc77a80, bamres=..., Seq=0)
    at D:/Sources/gemrb/gemrb/core/Projectile.cpp:92
#3  0x00007ff8f374a170 in GemRB::Projectile::Setup (this=0x25babc77a80)
    at D:/Sources/gemrb/gemrb/core/Projectile.cpp:319
#4  0x00007ff8f374cfb5 in GemRB::Projectile::Update (this=0x25babc77a80)
    at D:/Sources/gemrb/gemrb/core/Projectile.cpp:1220
#5  0x00007ff8f3734ddf in GemRB::Map::DrawMap (this=0x25ba742a1c0,
    viewport=..., dFlags=0) at D:/Sources/gemrb/gemrb/core/Map.cpp:1523
#6  0x00007ff8f36a78bc in GemRB::GameControl::DrawSelf (this=0x25b9f605c50,
    screen=...) at D:/Sources/gemrb/gemrb/core/GUI/GameControl.cpp:551
#7  0x00007ff8f36c3a90 in GemRB::View::Draw (this=0x25b9f605c50)
    at D:/Sources/gemrb/gemrb/core/GUI/View.cpp:266
#8  0x00007ff8f36c3462 in GemRB::View::DrawSubviews (this=0x25b9574b7d0)
    at D:/Sources/gemrb/gemrb/core/GUI/View.cpp:201
#9  0x00007ff8f36c3b34 in GemRB::View::Draw (this=0x25b9574b7d0)
    at D:/Sources/gemrb/gemrb/core/GUI/View.cpp:277
#10 0x00007ff8f36c3462 in GemRB::View::DrawSubviews (this=0x25b9574b6b0)
    at D:/Sources/gemrb/gemrb/core/GUI/View.cpp:201
#11 0x00007ff8f36c3b34 in GemRB::View::Draw (this=0x25b9574b6b0)
    at D:/Sources/gemrb/gemrb/core/GUI/View.cpp:277
#12 0x00007ff8f36ca54e in GemRB::WindowManager::DrawWindows (
    this=0x25b9522eb80)
    at D:/Sources/gemrb/gemrb/core/GUI/WindowManager.cpp:580
#13 0x00007ff8f37160ea in GemRB::Interface::Main (this=0x25b922876e0)
    at D:/Sources/gemrb/gemrb/core/Interface.cpp:646
#14 0x00007ff7f41d16ba in main (argc=1, argv=0x25b922a2750)
    at D:/Sources/gemrb/platforms/windows/GemRB.cpp:53
```

**Comment by bradallred** [comment]
is `Cycle` out of bounds?

**Comment by bradallred** [comment]
might be coincidence, but `Animation::MirrorAnimation` makes me feel like this could be related to #1536

**Comment by bradallred** [comment]
I dont know why that last commit isn't showing here; its definitely [there](https://github.com/gemrb/gemrb/tree/EffectQueue%2BProjectile_copy_issues)

**Comment by MarcelHB** [comment]
Github [reports ](https://www.githubstatus.com/)minor trouble.

Anyway, the previous trigger is fixed. Now I threw a fireball oil and ended up in something segfaulting:
```
#5  0x00007ff952d59d40 in msvcrt!malloc () from C:\Windows\System32\msvcrt.dll
#6  0x00007ff911f3fd8c in ?? () from C:\MSYS2\mingw64\bin\libstdc++-6.dll
#7  0x00007ff8f38258ea in __gnu_cxx::new_allocator<GemRB::Holder<GemRB::Sprite2D> >::allocate (this=0x8b937fe5c0, __n=12)
    at C:/MSYS2/mingw64/include/c++/11.2.0/ext/new_allocator.h:127
#8  0x00007ff8f388b7e6 in std::allocator_traits<std::allocator<GemRB::Holder<GemRB::Sprite2D> > >::allocate (__a=..., __n=12)
    at C:/MSYS2/mingw64/include/c++/11.2.0/bits/alloc_traits.h:460
#9  0x00007ff8f3872f17 in std::_Vector_base<GemRB::Holder<GemRB::Sprite2D>, std::allocator<GemRB::Holder<GemRB::Sprite2D> > >::_M_allocate (
    this=0x8b937fe5c0, __n=12)
    at C:/MSYS2/mingw64/include/c++/11.2.0/bits/stl_vector.h:346
#10 0x00007ff8f38bd3f4 in std::vector<GemRB::Holder<GemRB::Sprite2D>, std::allocator<GemRB::Holder<GemRB::Sprite2D> > >::reserve (this=0x8b937fe5c0, __n=12)
    at C:/MSYS2/mingw64/include/c++/11.2.0/bits/vector.tcc:78
#11 0x00007ff8f3683c62 in GemRB::AnimationFactory::GetCycle (
    this=0x1d34bcc6150, cycle=0)
    at D:/Sources/gemrb/gemrb/core/AnimationFactory.cpp:50
#12 0x00007ff8f37499b2 in GemRB::Projectile::CreateOrientedAnimations (
    this=0x1d34baa63b0, af=0x1d34bcc6150, Seq=0)
    at D:/Sources/gemrb/gemrb/core/Projectile.cpp:154
#13 0x00007ff8f3749743 in GemRB::Projectile::CreateAnimations (
    this=0x1d34baa63b0, bamres=..., Seq=0)
    at D:/Sources/gemrb/gemrb/core/Projectile.cpp:92
#14 0x00007ff8f374a24e in GemRB::Projectile::Setup (this=0x1d34baa63b0)
    at D:/Sources/gemrb/gemrb/core/Projectile.cpp:319
#15 0x00007ff8f374d093 in GemRB::Projectile::Update (this=0x1d34baa63b0)
    at D:/Sources/gemrb/gemrb/core/Projectile.cpp:1220
#16 0x00007ff8f3734e7f in GemRB::Map::DrawMap (this=0x1d347250810,
    viewport=..., dFlags=0) at D:/Sources/gemrb/gemrb/core/Map.cpp:1523
```
I can later try to rerun this on Valgrind.

**Comment by MarcelHB** [comment]
Valgrind report 1:
```
==4945== Invalid read of size 8
==4945==    at 0x17AF12: GemRB::Animation::MirrorAnimation(GemRB::BlitFlags) (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2EAA8F: GemRB::Projectile::CreateOrientedAnimations(GemRB::AnimationFactory const*, int) const (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2ECB30: GemRB::Projectile::CreateAnimations(GemRB::FixedSizeString<8ul, &strncasecmp> const&, int) (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2ECE6C: GemRB::Projectile::Setup() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2EE56C: GemRB::Projectile::Update() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2CD741: GemRB::Map::DrawMap(GemRB::Region const&, unsigned int) (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x1D7D1F: GemRB::GameControl::DrawSelf(GemRB::Region const&, GemRB::Region const&) (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2063F2: GemRB::View::Draw() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x206433: GemRB::View::Draw() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x206433: GemRB::View::Draw() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x20C796: GemRB::WindowManager::DrawWindows() const [clone .part.0] (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x29290A: GemRB::Interface::Main() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==  Address 0x25342788 is 72 bytes inside a block of size 104 free'd
==4945==    at 0x484684F: operator delete(void*) (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==4945==    by 0x2EAA6E: GemRB::Projectile::CreateOrientedAnimations(GemRB::AnimationFactory const*, int) const (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2ECB30: GemRB::Projectile::CreateAnimations(GemRB::FixedSizeString<8ul, &strncasecmp> const&, int) (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2ECE6C: GemRB::Projectile::Setup() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2EE56C: GemRB::Projectile::Update() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2CD741: GemRB::Map::DrawMap(GemRB::Region const&, unsigned int) (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x1D7D1F: GemRB::GameControl::DrawSelf(GemRB::Region const&, GemRB::Region const&) (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2063F2: GemRB::View::Draw() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x206433: GemRB::View::Draw() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x206433: GemRB::View::Draw() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x20C796: GemRB::WindowManager::DrawWindows() const [clone .part.0] (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x29290A: GemRB::Interface::Main() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==  Block was alloc'd at
==4945==    at 0x4843FB3: operator new(unsigned long) (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==4945==    by 0x17BE39: GemRB::AnimationFactory::GetCycle(unsigned short) const (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2EA930: GemRB::Projectile::CreateOrientedAnimations(GemRB::AnimationFactory const*, int) const (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2ECB30: GemRB::Projectile::CreateAnimations(GemRB::FixedSizeString<8ul, &strncasecmp> const&, int) (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2ECE6C: GemRB::Projectile::Setup() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2EE56C: GemRB::Projectile::Update() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2CD741: GemRB::Map::DrawMap(GemRB::Region const&, unsigned int) (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x1D7D1F: GemRB::GameControl::DrawSelf(GemRB::Region const&, GemRB::Region const&) (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2063F2: GemRB::View::Draw() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x206433: GemRB::View::Draw() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x206433: GemRB::View::Draw() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x20C796: GemRB::WindowManager::DrawWindows() const [clone .part.0] (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
```
Valgrind report 2:
```
==4945== Conditional jump or move depends on uninitialised value(s)
==4945==    at 0x2EE366: GemRB::Projectile::DoStep(unsigned int) (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2EE55B: GemRB::Projectile::Update() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2EF6E0: GemRB::Projectile::Draw(GemRB::Region const&) (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2CDA66: GemRB::Map::DrawMap(GemRB::Region const&, unsigned int) (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x1D7D1F: GemRB::GameControl::DrawSelf(GemRB::Region const&, GemRB::Region const&) (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2063F2: GemRB::View::Draw() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x206433: GemRB::View::Draw() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x206433: GemRB::View::Draw() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x20C796: GemRB::WindowManager::DrawWindows() const [clone .part.0] (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x29290A: GemRB::Interface::Main() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x1552FA: main (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
```
Number 3 (claimed to be the fatal one):
```
==4945==  Access not within mapped region at address 0x50
==4945==    at 0x2EB7BB: GemRB::Projectile::DrawTravel(GemRB::Region const&) (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2EF6C3: GemRB::Projectile::Draw(GemRB::Region const&) (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2CDA66: GemRB::Map::DrawMap(GemRB::Region const&, unsigned int) (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x1D7D1F: GemRB::GameControl::DrawSelf(GemRB::Region const&, GemRB::Region const&) (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x2063F2: GemRB::View::Draw() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x206433: GemRB::View::Draw() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x206433: GemRB::View::Draw() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x20C796: GemRB::WindowManager::DrawWindows() const [clone .part.0] (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x29290A: GemRB::Interface::Main() (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
==4945==    by 0x1552FA: main (in /mnt/shared/Sources/gemrb/build_linux/gemrb/gemrb)
```

**Comment by bradallred** [comment]
ug, the `PathNode` pointers are obviously not going to work with default copy/move

**Comment by MarcelHB** [comment]
We need to supply this on our own then. Not sure if a `PathNode` can be part of a circle? That's the only thing to take care of I think.

**Comment by lynxlynxlynx** [comment]
We should get rid of it eventually, there's even an issue reported for that: #880

**Comment by bradallred** [comment]
ok, that resolves the use-after-free for the Projectile path. something still seems broken tho.

**Comment by MarcelHB** [comment]
Can we simply co-move that iterator `step`? I can't find anything conclusive but to me it feels like it keeps the reference to children that have been deleted after everything else has moved (line 1554f).

**Comment by bradallred** [comment]
no, it's not possible to have an iterator member and move/copy it in a well behaved way. I changed it to an index and vector instead of list.

**Comment by MarcelHB** [comment]
Great, no more crashing.

But I found at least two disruptions compared to master:
1. I observed that some effects don't stop making noise as if they accumulate sound playback loops for much longer time, for example the molotows in IWD2.
2. When casting a fireball, the explosion waves have disappeared, and also so the cloud spells have lost a lot of VFX.

**Comment by bradallred** [comment]
I'll have to rewrite the history a bit so that it is stable enough to bisect

**Comment by bradallred** [comment]
ok, I cant do much more until sometime tomorrow. let me know if that helps reveal which commit broke things

**Comment by MarcelHB** [comment]
e321fc7444140e157cfcfa1d95f41a65543f3ce2 is when the effects no longer play in order. fcdff7ca9bc8063c7dc22db6381b1c275c2e0d5f is when the sound effects start to be odd.

**Comment by bradallred** [comment]
I did spot issues with e321fc7 and pushed fixes for those.

Annoyingly, I couldn't spot anything wrong with b5a4ec5. 90% of that is just adding/removing `const`, changing `->` to `.`, and adding `&` to `returns`. 90% of the remainder is just duplicate code to work with const `EffectQueue`s. I still went back and reversed some earlier attempts at avoiding `const_cast` to make the diff easier to sift though for meaningful differences.

File: gemrb/core/EffectQueue.cpp [hunk 2]
```diff
@@ -469,22 +460,19 @@ bool EffectQueue::RemoveEffect(const Effect* fx)
 void EffectQueue::ApplyAllEffects(Actor* target) const
 {
 	for (auto fx : effects) {
```
Comment (MarcelHB) [inline gemrb/core/EffectQueue.cpp:448]: W.r.t. the sound effects hanging around to much, I suspect the one place where we have a copy of the effect. I found that the effect that is being assigned the `FX_DURATION_JUST_EXPIRED` is not the same as the one from the list.

Comment (MarcelHB) [inline gemrb/core/EffectQueue.cpp:448]: W.r.t. the sound effects hanging around to much, I suspect this to be one place where we have a copy of the effect. I found that the effect that is being assigned the `FX_DURATION_JUST_EXPIRED` is not the same as the one from the list.

Comment (MarcelHB) [inline gemrb/core/EffectQueue.cpp:448]: W.r.t. the sound effects hanging around to much, I suspect this to be one place where we have a copy of the effect. I found that the effect that is being assigned the `FX_DURATION_JUST_EXPIRED` is not the same as the one from the list.

Comment (MarcelHB) [inline gemrb/core/EffectQueue.cpp:448]: W.r.t. the sound effects hanging around to much, I suspect this to be one place where we have a copy of the effect. I found that the effect that is being assigned the `FX_DURATION_JUST_EXPIRED` is not the same as the one from the list.

Comment (MarcelHB) [inline gemrb/core/EffectQueue.cpp:448]: W.r.t. the sound effects hanging around to much, I suspect this to be one place where we have a copy of the effect. I found that the effect that is being assigned the `FX_DURATION_JUST_EXPIRED` is not the same as the one from the list.

**Comment by MarcelHB** [comment]
Grr, GitHub returns 500 codes r/n when commenting inline, so:

https://github.com/gemrb/gemrb/blob/7f69009df22e817366f2fe7079360b02328f7ac0/gemrb/core/EffectQueue.cpp#L448

W.r.t. the sound effects hanging around to much, I suspect this to be one place where we have a copy of the effect. I found that the effect that is being assigned the `FX_DURATION_JUST_EXPIRED` is not the same as the one from the list. Previously, with the pointers list, this was no problem.

**Comment by bradallred** [comment]
yup, we stumbled across some bad iterator code thats been there forever. should be fixed in the latest commit.

**Comment by MarcelHB** [comment]
The cloud and fireballs are still incomplete but I'm not sure if there are two problems and if so, if the later one (no more secondary projectiles at all) supersedes the other one (timing/direction of projectiles feels bad, cannot confirm yet). Maybe I can get another round of bisection later.

**Comment by bradallred** [comment]
the obvious suspect is 297aba7

**Comment by lynxlynxlynx** [comment]
Shouldn't all travel/shadow[0] checks be now converted as well? Maybe it's harmless in some cases, but definitely less clear.

**Comment by bradallred** [comment]
> Shouldn't all travel/shadow[0] checks be now converted as well? Maybe it's harmless in some cases, but definitely less clear.

How do you mean? `Animation` has an `operator bool()` so that empty animations are false. Such animations used to be `nullptr`, so logically they should be the same. Is there a specific instance in question?

**Comment by lynxlynxlynx** [comment]
Ah, ok. Limits of phone review. :)

**Comment by MarcelHB** [comment]
Things look better, but to give you an impression what still bothers me:

master
iwd2_fb_master

this branch
iwd2_fb

**Comment by lynxlynxlynx** [comment]
Looks like as if the spread animation went through all its phases immediately. Might  be fruitful to step through Payload or ChangePhase while keeping game time in check.

**Comment by bradallred** [comment]
it could also be a problem with `GetLinePath`

**Comment by bradallred** [comment]
yeah, I was _prepending_ new path nodes instead of appending

**Comment by MarcelHB** [comment]
:+1: Anyway I found something else: When casting a very specific IWD2 cleric spell (_Flame strike_, sppr503), the game crashes:
```
Thread 1 "gemrb" received signal SIGABRT, Aborted.
__pthread_kill_implementation (no_tid=0, signo=6, threadid=140737315291584) at pthread_kill.c:44
44      pthread_kill.c: No such file or directory.
(gdb) bt
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=140737315291584) at pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=140737315291584) at pthread_kill.c:80
#2  __GI___pthread_kill (threadid=140737315291584, signo=signo@entry=6) at pthread_kill.c:91
#3  0x00007ffff70d4476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x00007ffff70ba7b7 in __GI_abort () at abort.c:79
#5  0x00007ffff711b5e6 in __libc_message (action=action@entry=do_abort, fmt=fmt@entry=0x7ffff726d13d "%s
") at ../sysdeps/posix/libc_fatal.c:155
#6  0x00007ffff7132adc in malloc_printerr (str=str@entry=0x7ffff7270330 "malloc(): unaligned tcache chunk detected") at malloc.c:5543
#7  0x00007ffff7136c8c in tcache_get (tc_idx=<optimized out>) at malloc.c:3082
#8  __GI___libc_malloc (bytes=100) at malloc.c:3200
#9  0x00007ffff7463cbc in operator new(unsigned long) () at /lib/x86_64-linux-gnu/libstdc++.so.6
#10 0x00007ffff75147cc in std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >::_M_mutate(unsigned long, unsigned long, wchar_t const*, unsigned long) () at /lib/x86_64-linux-gnu/libstdc++.so.6
#11 0x00007ffff75150b5 in std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >::push_back(wchar_t) () at /lib/x86_64-linux-gnu/libstdc++.so.6
#12 0x0000555555938169 in GemRB::TLKImporter::ResolveTags(std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&) (this=0x5555564e3f20, source=L"Erleidet <AMOUNT> Schaden durch <TYPE> von <DAMAGER>")
    at /mnt/shared/Sources/gemrb/gemrb/plugins/TLKImporter/TLKImporter.cpp:326
#13 0x000055555593869f in GemRB::TLKImporter::GetString[abi:cxx11](GemRB::ieStrRef, GemRB::STRING_FLAGS) (this=0x5555564e3f20, strref=25028, flags=(GemRB::STRING_FLAGS::SOUND | GemRB::STRING_FLAGS::SPEECH | GemRB::STRING_FLAGS::RESOLVE_TAGS))
    at /mnt/shared/Sources/gemrb/gemrb/plugins/TLKImporter/TLKImporter.cpp:385
#14 0x000055555570f870 in GemRB::Interface::GetString[abi:cxx11](GemRB::ieStrRef, GemRB::STRING_FLAGS) const (this=0x555555be7fd0, strref=25028, options=(GemRB::STRING_FLAGS::SOUND | GemRB::STRING_FLAGS::SPEECH | GemRB::STRING_FLAGS::RESOLVE_TAGS))
    at /mnt/shared/Sources/gemrb/gemrb/core/Interface.cpp:1678
#15 0x00005555555da9e2 in GemRB::DisplayMessage::DisplayConstantStringName(unsigned long, GemRB::Color const&, GemRB::Scriptable const*) const (this=0x55555675a950, stridx=196, color=..., speaker=0x555556c81ca0)
    at /mnt/shared/Sources/gemrb/gemrb/core/DisplayMessage.cpp:224
#16 0x00005555557c6a3a in GemRB::Actor::DisplayCombatFeedback(unsigned int, int, int, GemRB::Scriptable const*) (this=0x555556c81ca0, damage=49, resisted=0, damagetype=8, hitter=0x555556c81ca0) at /mnt/shared/Sources/gemrb/gemrb/core/Scriptable/Actor.cpp:4658
#17 0x00005555557c60a0 in GemRB::Actor::Damage(int, int, GemRB::Scriptable*, int, int, int) (this=0x555556c81ca0, damage=49, damagetype=8, hitter=0x555556c81ca0, modtype=0, critical=0, saveflags=8) at /mnt/shared/Sources/gemrb/gemrb/core/Scriptable/Actor.cpp:4497
#18 0x000055555589e059 in fx_damage(GemRB::Scriptable*, GemRB::Actor*, GemRB::Effect*) (target=0x555556c81ca0, fx=0x55555968abf0) at /mnt/shared/Sources/gemrb/gemrb/plugins/FXOpcodes/FXOpcodes.cpp:1395
#19 0x00005555555ffad0 in GemRB::EffectDesc::operator()(GemRB::Scriptable*, GemRB::Actor*, GemRB::Effect*) const (this=0x555555b34fa0 <GemRB::Opcodes+288>, s=0x555556c81ca0, a=0x555556c81ca0, fx=0x55555968abf0) at /mnt/shared/Sources/gemrb/gemrb/core/EffectQueue.h:182
#20 0x00005555555fbc56 in GemRB::EffectQueue::ApplyEffect(GemRB::Actor*, GemRB::Effect*, unsigned int, unsigned int) const (this=0x555559c6e6a0, target=0x555556c81ca0, fx=0x55555968abf0, first_apply=1, resistance=1)
    at /mnt/shared/Sources/gemrb/gemrb/core/EffectQueue.cpp:1239
#21 0x00005555555f9baf in GemRB::EffectQueue::AddEffect(GemRB::Effect*, GemRB::Scriptable*, GemRB::Actor*, GemRB::Point const&) const (this=0x555559c6e6a0, fx=0x55555968abf0, self=0x555556c81ca0, pretarget=0x555556c81ca0, dest=...)
    at /mnt/shared/Sources/gemrb/gemrb/core/EffectQueue.cpp:608
#22 0x00005555555fa185 in GemRB::EffectQueue::AddAllEffects(GemRB::Actor*, GemRB::Point const&) (this=0x555559c6e6a0, target=0x555556c81ca0, destination=...) at /mnt/shared/Sources/gemrb/gemrb/core/EffectQueue.cpp:708
#23 0x0000555555798fc5 in GemRB::Projectile::Payload() (this=0x555559c6e510) at /mnt/shared/Sources/gemrb/gemrb/core/Projectile.cpp:551
...
```
This slighty confusing message above (`malloc(): unaligned tcache chunk detected`) is probably the reason of some memory error, as Valgrind reports:
```
==14606== Invalid write of size 1
==14606==    at 0x34B82F: GemRB::Projectile::CreateCompositeAnimation(GemRB::AnimationFactory const*, int) const (Projectile.cpp:104)
==14606==    by 0x34B6BF: GemRB::Projectile::CreateAnimations(GemRB::FixedSizeString<8ul, &strncasecmp> const&, int) (Projectile.cpp:80)
==14606==    by 0x34C31B: GemRB::Projectile::Setup() (Projectile.cpp:310)
==14606==    by 0x34F5BB: GemRB::Projectile::Update() (Projectile.cpp:1211)
==14606==    by 0x309B00: GemRB::Map::DrawMap(GemRB::Region const&, unsigned int) (Map.cpp:1523)
...
==14606==  Address 0x258ae34e is 14 bytes inside a block of size 104 free'd
==14606==    at 0x484684F: operator delete(void*) (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==14606==    by 0x34B80B: GemRB::Projectile::CreateCompositeAnimation(GemRB::AnimationFactory const*, int) const (Projectile.cpp:97)
==14606==    by 0x34B6BF: GemRB::Projectile::CreateAnimations(GemRB::FixedSizeString<8ul, &strncasecmp> const&, int) (Projectile.cpp:80)
==14606==    by 0x34C31B: GemRB::Projectile::Setup() (Projectile.cpp:310)
==14606==    by 0x34F5BB: GemRB::Projectile::Update() (Projectile.cpp:1211)
==14606==    by 0x309B00: GemRB::Map::DrawMap(GemRB::Region const&, unsigned int) (Map.cpp:1523)
...
==14606==  Block was alloc'd at
==14606==    at 0x4843FB3: operator new(unsigned long) (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==14606==    by 0x1717E3: GemRB::AnimationFactory::GetCycle(unsigned short) const (AnimationFactory.cpp:55)
==14606==    by 0x34B7B8: GemRB::Projectile::CreateCompositeAnimation(GemRB::AnimationFactory const*, int) const (Projectile.cpp:93)
==14606==    by 0x34B6BF: GemRB::Projectile::CreateAnimations(GemRB::FixedSizeString<8ul, &strncasecmp> const&, int) (Projectile.cpp:80)
==14606==    by 0x34C31B: GemRB::Projectile::Setup() (Projectile.cpp:310)
==14606==    by 0x34F5BB: GemRB::Projectile::Update() (Projectile.cpp:1211)
==14606==    by 0x309B00: GemRB::Map::DrawMap(GemRB::Region const&, unsigned int) (Map.cpp:1523)
...
``

**Comment by bradallred** [comment]
those two things look unrelated and the first is certainly not caused by changes here.

That first report seems fishy because TLKImporter.cpp:326 is `dest.push_back(srcch);` everything here is not our code. I don't know what to make of "malloc(): unaligned tcache chunk detected". never seen that before.

**Comment by MarcelHB** [comment]
Ok, all the stuff I tested for making things fly around and take effect look good now.

**Comment by bradallred** [comment]
did that resolve the flame strike issue then? if not let's open an issue for that.

**Comment by bradallred** [comment]
oh, and thanks for all the help testing!

**Comment by MarcelHB** [comment]
> did that resolve the flame strike issue then? if not let's open an issue for that.

Yep, that's fine now.

**Comment by bradallred** [comment]
I'm gonna merge this to get things building again. Its obvious there aren't major issues.

**Pull request merged**